### PR TITLE
Feature/docker refactor

### LINKED ISSUE #22: Docker Refactoring
These are a few things I have noted that could use work as I was working with the containers. Feel free to add more.

1. Lightweight containers should work well from bash with no default terminator/x11 forwarding
2. Too many mountpoints! It's easier to maintain and scale the fewer there are. Can copy most files/scripts needed into container builds instead.
3. In general the dockerfiles and compose config feels too complicated to work with/maintain reliably. Need to simplify more.
4. All containers should have `apt-get install iputils-ping net-tools` for network debug
5. Containers should be based off of official ROS images instead of Ubuntu. This would make them a bit cleaner.
6. Lightweight: only install the minimum libraries needed for HARMONI core. If you need more, make a variant container or put it into dev containers (dev containers can be a mess as long as they don't require much maintenance).

### PR DESCRIPTION
**Pull request by chrismbirmingham**
The goal of this branch is to simplify and reorganize the dockerfiles for faster builds. This should include a base package that Harmoni is built on. The docker-compose yamls also need to be simplified and reorganized with minimal mount points.

Finalizing the autobuild and additional basic testing may also get swept into this feature.

### CODE REVIEW

File: docker-compose-core.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      - ../HARMONI/:/root/local_mount/HARMONI/
+      # Configuration
+      - ~/.aws:/root/.aws/
+      # Other
+      - /tmp/.X11-unix:/tmp/.X11-unix
+      - /etc/timezone:/etc/timezone:ro
+      - /etc/localtime:/etc/localtime:ro
+    working_dir: /root/harmoni_catkin_ws/src/HARMONI
+    command:
+      tail -F anything
```
Comment (RMichaelSwan) [inline docker-compose-core.yml]: This will give an error message but technically still gives us the functionality we want (to keep the container running). Because of the message ("cannot open 'anything' for reading: No such file or directory"), I prefer `tail -f /dev/null`

File: dockerfiles/dev/harmoni/dockerfile [hunk 2]
```diff
@@ -1,88 +1,99 @@
-FROM cmbirmingham/ros-kinetic-dev:latest
-
+#!/bin/bash
+FROM harmoniteam/lightweight:harmoni
```
Comment (RMichaelSwan) [inline dockerfiles/dev/harmoni/dockerfile]:
```suggestion
FROM harmoniteam/lightweight:harmoni
```

File: docker-compose-hardware.yml [hunk 3]
```diff
@@ -5,55 +5,50 @@ services:
     container_name: harmoni_hardware
```
Comment (RMichaelSwan) [inline docker-compose-hardware.yml]: Should probably delete this file instead of lightweight-multipc.  docker-compose-harmoni-lightweight-multipc.yml is what I'm using for the "manual" multi-pc instructions right now (and TEST-docker-compose-harmoni-swarm.yml for automated multipc swarm that needs hostname work).  docker-compose-harmoni-hardware.yml is a precursor that sort of works but doesn't provide the needed networking.

File: docker-compose-harmoni-lightweight-multipc.yml [hunk 4]
```diff
@@ -1,212 +0,0 @@
-version: "3.7"
```
Comment (RMichaelSwan) [inline docker-compose-harmoni-lightweight-multipc.yml]: This file shouldn't be deleted. Primarily due to the network settings. See docker-compose-hardware comment.

File: dockerfiles/README.md [hunk 5]
```diff
@@ -1,37 +1,34 @@
...
          - _Note: you can also use the alias ```rlspeech``` we have provided_
 
 ## Why Docker?
 Although Harmoni will work without Docker, Harmoni is intended to be used with Docker to maximize portability and scalability. By using Docker Harmoni is quick and easy to set up on any OS or Hardware which currently supports docker.  We provide pre-built images for development, lightweight images for deployment, and even support for deploying to ARM chipsets.
 
-## Docker Containers
-We provide two sets of docker images. The first set is a heavier development set of images, which are based off of a ubuntu-16 image and contain graphical tools for use in a typical development cycle. These are intended to be helpful for developing interactively on a new machine. We also provide a set of lightweight images for deployment to space constrained machines, as well as images built with the arm architecture. 
+Docker also allows us to mix and match OS/ROS versions in different containers. For instance we use noetic with vision containers because of the python3-opencv integration but we use kinetic with w2l because facebook doesn't support w2l on 20.04 yet.
```
Comment (RMichaelSwan) [inline dockerfiles/README.md]: I know what I said before regarding messages, between ROS versions, but upon further review, I realize I'm not 100% on this. See discussion [here](https://answers.ros.org/question/360007/compatibility-of-ros-melodic-with-ros-kinetic/?answer=360034#post-id-360034) (especially answer and first comment). It appears to be a common discussion item that often leads to the answer "It's not meant to be compatible, but often it works fine" Seems the message format of dependencies may change between versions (not always, but could happen). As long as we're using our own message types to communicate between nodes, I don't see a problem, but if we're relying on something else, maybe we'd see issues.

TLDR: probably should weaken this strong statement a bit.

File: dockerfiles/noetic/base/dockerfile [hunk 6]
```diff
@@ -0,0 +1,126 @@
+#!/bin/bash
+FROM ros:noetic-perception-focal
```
Comment (RMichaelSwan) [inline dockerfiles/noetic/base/dockerfile]: Not to be pedantic, but how come you did perception here but base for the kinetic dockerfile?

File: harmoni_sensors/harmoni_camera/scripts/harmoni_camera/camera_service.py [hunk 7]
```diff
@@ -12,8 +12,8 @@
 # Other Imports
 import sys
 
-sys.path.remove("/opt/ros/kinetic/lib/python2.7/dist-packages")
-sys.path.append("/opt/ros/kinetic/lib/python2.7/dist-packages")
+# sys.path.remove("/opt/ros/kinetic/lib/python2.7/dist-packages")
+sys.path.append("/opt/ros/noetic/lib/python2.7/dist-packages")
```
Comment (RMichaelSwan) [inline harmoni_sensors/harmoni_camera/scripts/harmoni_camera/camera_service.py]: Does noetic have python2.7 packages? I thought it was python3 only.

**Review by RMichaelSwan — CHANGES_REQUESTED** [review]

File: dockerfiles/noetic/base/dockerfile (hunk 6 above)
Comment (chrismbirmingham) [reply dockerfiles/noetic/base/dockerfile]: This was because the perception contains opencv so we don't need to build it ourselves in noetic. In kinetic we are using python3 so the installed version of opencv won't work. I am mostly convinced we need to build ROS from scratch here if we are using python3 so starting from a ros dockerfile is likely not going to work anyway.

File: harmoni_sensors/harmoni_camera/scripts/harmoni_camera/camera_service.py (hunk 7 above)
Comment (chrismbirmingham) [reply harmoni_sensors/harmoni_camera/scripts/harmoni_camera/camera_service.py]: It is, this was leftover from some tinkering and should just be removed for noetic. But that does raise questions on how to handle the kinetic build...

File: dockerfiles/dev/harmoni/dockerfile (hunk 2 above)
Comment (chrismbirmingham) [reply dockerfiles/dev/harmoni/dockerfile]: I'm assuming this should be for all the dockerfiles?

File: docker-compose-harmoni-lightweight-multipc.yml (hunk 4 above)
Comment (chrismbirmingham) [reply docker-compose-harmoni-lightweight-multipc.yml]: Undeleted and renamed: docker-compose-multipc

**Comment by chrismbirmingham** [comment]
Just as a note - the images are 3.5-4 GB right now, a future refactor will be needed to further lighten them up. My primary concern right now is getting them working.

**Review by micolspitale93 — APPROVED** [review]
I think that the new structure and the renaming are now more clear.

**Comment by RMichaelSwan** [comment]
need to merge develop into this branch and fix conflicts. Can create a separate buildx yaml file if you'd like to do something custom.

File: docker-compose-core.yml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  harmoni_core:
+    container_name: harmoni_core
+    build:
+      context: .
+      dockerfile: dockerfiles/harmoni/$ROS_DISTRO/core/dockerfile
+      network: host
+    image: harmoniteam/harmoni:$ROS_DISTRO-core
+    init: true
+    environment:
+      DISPLAY: $DISPLAY
```
Comment (RMichaelSwan) [inline docker-compose-core.yml]: Should leave this out from core compose (no display needed)

File: docker-compose-core.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      - /etc/timezone:/etc/timezone:ro
+      - /etc/localtime:/etc/localtime:ro
+    working_dir: /root/harmoni_catkin_ws/src/HARMONI
+    command:
+      tail  -f /dev/null
+      #  bash -c "terminator -ue \"echo 'Entering harmoni_core Container... \\n start with roscore and rlharmoniservices' && bash\""
+
+networks:
+  ros_net:
+    driver: bridge
```
Comment (RMichaelSwan) [inline docker-compose-core.yml]: FYI, with this driver, the host won't be able to communicate with container apps except through the ports noted above ^

File: docker-compose.yml [hunk 10]
```diff
@@ -74,92 +64,73 @@ services:
...
     devices:
       - /dev/dri:/dev/dri
-      #- /dev/snd:/dev/snd
       - /dev/video0:/dev/video0
     volumes:
       - ../HARMONI/:/root/local_mount/HARMONI/
       # Configuration
-      - ./dockerfiles/config/setup_script.sh:/root/.setup_script.sh
-      - ./dockerfiles/config/dev_setup_script.sh:/root/.dev_setup_script.sh
+      - ~/.aws:/root/.aws/
```
Comment (RMichaelSwan) [inline docker-compose.yml:114]: I think there's a better way to inject secrets into containers, but this works for now. Might be good to make an issue to look into it later.

File: dockerfiles/README.md [hunk 11]
```diff
@@ -1,37 +1,34 @@
...
 
-## Docker Containers
-We provide two sets of docker images. The first set is a heavier development set of images, which are based off of a ubuntu-16 image and contain graphical tools for use in a typical development cycle. These are intended to be helpful for developing interactively on a new machine. We also provide a set of lightweight images for deployment to space constrained machines, as well as images built with the arm architecture. 
+Docker also allows us to (mostly) mix and match OS/ROS versions in different containers. For instance we use noetic with vision containers because of the python3-opencv integration but we use kinetic with w2l because facebook doesn't support w2l on 20.04 yet. This works for communication with standard messages and harmoni messages but may cause issues with other message types. [See here for more information.]()
```
Comment (RMichaelSwan) [inline dockerfiles/README.md]:
```suggestion
Docker also allows us to (mostly) mix and match OS/ROS versions in different containers. For instance we use noetic with vision containers because of the python3-opencv integration but we use kinetic with w2l because facebook doesn't support w2l on 20.04 yet. This works for communication with standard messages and harmoni messages but may cause issues with other message types. [See here for more information.](https://answers.ros.org/question/45153/ros-fuerte-node-communicating-with-ros-electric-node/?answer=45160#post-id-45160)
```
One possible link. There are a bunch of similar questions on ROS answers if you look for them. They generally have the same answer (not supported, but often works)
